packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol

```diff
@@ -0,0 +1,208 @@
+        uint32  nonce;
+        uint16  tokenID;
+        uint    amount;
+        uint16  feeTokenID;
+        uint    fee;
+        address to;
+        bytes32 dataHash;
+        uint24  minGas;
+    }
+
+    // Auxiliary data for each withdrawal
+    struct WithdrawalAuxiliaryData
+    {
+        uint gasLimit;
+        bytes signature;
+        bytes auxiliaryData;
+    }
+
+    event ForcedWithdrawalConsumed(
+        uint24  indexed accountID,
+        uint16          tokenID,
+        uint            amount
+    );
+
+    function process(
+        ExchangeData.State storage S,
+        ExchangeData.BlockContext memory ctx,
+        bytes memory data,
+        bytes memory auxiliaryData
+        )
+        internal
+        returns (uint feeETH)
+    {
+        Withdrawal memory withdrawal = readWithdrawal(data);
+
+        WithdrawalAuxiliaryData memory auxData = abi.decode(auxiliaryData, (WithdrawalAuxiliaryData));
+
+        if (withdrawal.withdrawalType == 0) {
+            // Signature checked offchain, nothing to do
+        } else if (withdrawal.withdrawalType == 1) {
+            // Check appproval onchain
+
+            // Calculate the tx hash
+            bytes32 txHash = EIP712.hashPacked(
+                ctx.DOMAIN_SEPARATOR,
+                keccak256(
+                    abi.encode(
+                        WITHDRAWAL_TYPEHASH,
+                        withdrawal.owner,
+                        withdrawal.accountID,
+                        withdrawal.nonce,
+                        withdrawal.tokenID,
+                        withdrawal.amount,
+                        withdrawal.feeTokenID,
+                        withdrawal.fee,
+                        withdrawal.to,
+                        withdrawal.dataHash,
+                        withdrawal.minGas
+                    )
+                )
+            );
+
+            // Verify the signature if one is provided, otherwise fall back to an approved tx
+            if (auxData.signature.length > 0) {
+                require(txHash.verifySignature(withdrawal.owner, auxData.signature), "INVALID_SIGNATURE");
+            } else {
+                require(S.approvedTx[withdrawal.owner][txHash], "TX_NOT_APPROVED");
+                S.approvedTx[withdrawal.owner][txHash] = false;
+            }
+        } else if (withdrawal.withdrawalType == 2 || withdrawal.withdrawalType == 3) {
+            // Forced withdrawals cannot make use of certain features because the
+            // necessary data is not authorized by the account owner.
+            require(withdrawal.owner == withdrawal.to, "INVALID_WITHDRAWAL_ADDRESS");
+            require(withdrawal.fee == 0, "FEE_NOT_ZERO");
+            require(auxData.auxiliaryData.length == 0, "AUXILIARY_DATA_NOT_ALLOWED");
+
+            ExchangeData.ForcedWithdrawal storage forcedWithdrawal = S.pendingForcedWithdrawals[withdrawal.accountID][withdrawal.tokenID];
+            if (forcedWithdrawal.timestamp == 0) {
```

Comment:
We should make this type the 5th type: offchain operator initiated withdrawal without user authorization to make it clear.

Reply:
From other comment:

> The different types are for the different behaviours in the circuit. withdrawType = 4 like this would not behave differently than types 2/3 in the circuit (so we'd have to have 4/5, but they would do exactly the same thing).

> The types are not related to how the onchain verfication is done. Type 1 withdrawals can be authorized onchain with an ECDSA signature or an approved hash, but the actual behaviour is exactly the same, it's just authorized differently, so they use the same type. The operator just needs to provide the correct authentication.

Reply:
Alright, better documentation will do :)